File: java/Minimum Multiplications to reach End.java
class Solution {
    int minimumMultiplications(int[] arr, int start, int end) {

        // Your code here
        if(start==end) return 0;
        Queue<Pair> q = new LinkedList<>();
        int vis[] = new int[100000];
        q.add(new Pair(0,start));
        
        while(q.size()>0){
            Pair p = q.remove();
            int steps = p.a;
            int num = p.b;
            
            for(int i=0; i<arr.length; i++){
                long temp = arr[i]*num % 100000;
                int ns = (int)temp;
                if(ns==end) return steps+1;
                if(vis[ns]==0){
                    q.add(new Pair(steps+1,ns));
                    vis[ns] = 1;
                }
            }
        }
        
        return -1;
    }
}

class Pair{
    int a;
    int b;
    Pair(int a, int b){
        this.a = a;
        this.b = b;
    }
    
}

File: cpp/Minimum Multiplications to reach End.cpp
class Solution {
  public:
    int minimumMultiplications(vector<int>& arr, int start, int end) {
        // code here
        if(start==end) return 0;
        queue<pair<int,int>>q;
        q.push({start,0});
        vector<int>dist(100000,1e9);
        dist[start]=0;
        int mod=100000;
        while(!q.empty()){
            int node=q.front().first;            
            int steps=q.front().second;
            q.pop();
            for(auto it:arr){
                long num=(node*it)%mod;
                if(steps+1<dist[num]) {
                    dist[num]=steps+1;
                    q.push({num,steps+1});
                }
                if(num==end) return steps+1;
                
            }
        }
        return -1;
    }

};